iati/core/rulesets.py

```diff
@@ -1,45 +1,841 @@
+            return
+        # Checks that anything after the YYYY-MM-DD string is a permitted timezone character
+        pattern = re.compile(r'^([+-]([01][0-9]|2[0-3]):([0-5][0-9])|Z)?$')
+        if (len(set(dates)) == 1) and pattern.match(dates[0][10:]):
+            if len(dates[0]) < 10:
+                # '%d' and '%m' are documented as requiring zero-padded dates.as input. This is actually for output. As such, a separate length check is required to ensure zero-padded values.
+                raise ValueError
+            return datetime.strptime(dates[0][:10], '%Y-%m-%d')
+        raise ValueError
+
+    def _check_against_Rule(self, context_element):
+        """Assert that the date value of `less` is chronologically before the date value of `more`.
+
+        Args:
+            context_element (Element): An XML Element.
+
+        Return:
+            bool: Return `True` when `less` is chronologically before `more`.
+                  Return `False` when `less` is not chronologically before `more`.
+            None: When a condition is met to skip validation.
+
+        Raises:
+            ValueError: When a date is given that is not in the correct xsd:date format.
+
+        Note:
+            `date` restricted to 10 characters in order to exclude possible timezone values.
+
+        """
+        early_date = self._get_date(context_element, self.less)
+        later_date = self._get_date(context_element, self.more)
+
+        try:
+            # python2 allows `bool`s to be compared to `None` without raising a TypeError, while python3 does not
+            if early_date is None or later_date is None:
+                return None
+
+            if early_date >= later_date:
+                return False
+        except TypeError:
+            # a TypeError is raised in python3 if either of the dates is None
+            return None
+        return True
+
+
+class RuleDependent(Rule):
+    """Representation of a Rule that checks that if one of the Elements or Attributes in a given `path` exists then all its dependent Elements or Attributes must also exist."""
+
+    def __init__(self, context, case):
+        """Initialise a `dependent` rule."""
+        self.name = 'dependent'
+
+        super(RuleDependent, self).__init__(context, case)
+
+    def __str__(self):
+        """Return string stating what TestRuleDependent is checking."""
+        if len(self.paths) == 1:
+            return 'Within each `{self.context}`, either `{self.paths[0]}` exists or it does not. As such, this Rule is always True.'.format(**locals())
+        return 'Within each `{self.context}`, either none of `{0}` must exist, or they must all exist.'.format('` or `'.join(self.paths), **locals())
+
+    def _check_against_Rule(self, context_element):
+        """Assert that either all given `paths` or none of the given `paths` exist for the `context_element`.
+
+        Args:
+            context_element (Element): An XML Element.
```

Comment:
To be consistent with other parts of the library (and assuming it is this), should be:

* `context_element (etree._Element): An lxml element tree representing the provided XML.`

(here and elsewhere)

Reply:
done